sources/models/src/modeled_types/ecs.rs

```diff
@@ -0,0 +1,172 @@
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+// Just need serde's Error in scope to get its trait methods
+use serde::de::Error as _;
+use snafu::ensure;
+use std::borrow::Borrow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Deref;
+use super::error;
+
+/// ECSAttributeKey represents a string that contains a valid ECS attribute key.  It stores
+/// the original string and makes it accessible through standard traits.
+// https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Attribute.html
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ECSAttributeKey {
+    inner: String,
+}
+
+
+// The name of the attribute. Up to 128 letters (uppercase and lowercase), numbers, hyphens,
+// underscores, and periods are allowed.
+lazy_static! {
+    pub(crate) static ref ECS_ATTRIBUTE_KEY: Regex = Regex::new(
+        r"(?x)^
+       [a-zA-Z0-9-_\./]{1,128}
+   $"
```

Comment:
```suggestion
        r"(?x)^
          [a-zA-Z0-9_./-]{1,128}
          $"
```
* spacing suggestion
* you don't need to escape `.` in `[]`
* it's convention to put `-` last so someone doesn't think it's another range.

Reply:
> you don't need to escape `.` in `[]`

`.` isn't escaped, a `\` is a valid character in the attribute name.  Now that the documentation is updated, I'll update the comment above this.

Rewriting the character set to `[a-zA-Z0-9._/\-]`.

Reply:
`\` is a valid character, but you have to escape it, right?  So `\\`?  With `\-` you're escaping `-` which doesn't need escaping and so is the same as just writing `-`.

Reply:
Yeah, I had to change it to `[a-zA-Z0-9._/\\-]`; I wrote the previous comment before actually testing it.

Reply:
The bigger issue is that the datastore is preventing me from having a key with a backslash:

```
$ apiclient -u /settings -m PATCH -d '{"ecs":{"instance-attributes":{"foo\\/":"\\ /bar \\"}}}'
Status 500 when PATCHing unix://2f72756e2f6170692e736f636b:0/settings: Error serializing Settings: Error creating valid datastore key: Key name 'foo\/' has invalid format: Segment 'foo\/' contains invalid character '\' 
```

Reply:
I vote for banning evil backslashes.

Reply:
I'll remove them and add this to the README in ecs-agent where I'm tracking work.

Reply:
Done